provider/datastore/keystore.go

```diff
@@ -0,0 +1,521 @@
+		for _, h := range stored {
+			if _, ok := uniq[string(h)]; !ok {
+				uniq[string(h)] = struct{}{}
+				result = append(result, h)
+			}
+		}
+	}
+	return result, nil
+}
+
+// ContainsPrefix reports whether the KeyStore currently holds at least one
+// multihash whose kademlia identifier (bit256.Key) starts with the provided
+// bit-prefix.
+func (s *KeyStore) ContainsPrefix(ctx context.Context, prefix bitstr.Key) (bool, error) {
+	if s.closed() {
+		return false, ErrKeyStoreClosed
+	}
+	s.lk.Lock()
+	defer s.lk.Unlock()
+
+	// Case 1: shorter than bucket length — any bucket with this key-prefix is enough.
+	if len(prefix) < s.prefixLen {
+		rem := s.prefixLen - len(prefix)
+		limit := 1 << rem
+		for i := range limit {
+			suffix := fmt.Sprintf("%0*b", rem, i)
+			dsKey := s.dsKey(prefix + bitstr.Key(suffix))
+			exists, err := s.ds.Has(ctx, dsKey)
+			if err != nil {
+				return false, err
+			}
+			if exists {
+				return true, nil
+			}
+		}
+		return false, nil
+	}
+
+	// Case 2: at least a full bucket prefix — check that bucket's content.
+	dsKey := s.dsKey(bitstr.Key(prefix[:s.prefixLen]))
+
+	// Fast path when asking exactly for a bucket prefix: existence implies non-empty.
+	if len(prefix) == s.prefixLen {
+		exists, err := s.ds.Has(ctx, dsKey)
+		return exists, err
+	}
+
+	// Longer-than-bucket: must inspect entries and see if any starts with `prefix`.
+	data, err := s.ds.Get(ctx, dsKey)
+	if err != nil {
+		if err == ds.ErrNotFound {
+			return false, nil
+		}
+		return false, err
+	}
+	var stored []mh.Multihash
+	if err := json.Unmarshal(data, &stored); err != nil {
+		return false, err
+	}
+	for _, h := range stored {
+		if keyspace.IsPrefix(prefix, keyspace.MhToBit256(h)) {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// emptyLocked deletes all entries under the datastore prefix, assuming s.lk is
+// already held.
+func (s *KeyStore) emptyLocked(ctx context.Context) error {
+	res, err := s.ds.Query(ctx, query.Query{Prefix: s.base.String()})
+	if err != nil {
+		return err
+	}
+	defer res.Close()
+
+	for r := range res.Next() {
+		if r.Error != nil {
+			return r.Error
+		}
+		if err := s.ds.Delete(ctx, ds.NewKey(r.Key)); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Empty deletes all entries under the datastore prefix.
+func (s *KeyStore) Empty(ctx context.Context) error {
+	if s.closed() {
+		return ErrKeyStoreClosed
+	}
+	s.lk.Lock()
+	defer s.lk.Unlock()
+
+	return s.emptyLocked(ctx)
+}
+
+// Delete removes the given keys from datastore.
+func (s *KeyStore) Delete(ctx context.Context, keys ...mh.Multihash) error {
+	if len(keys) == 0 {
+		return nil
+	}
+	s.lk.Lock()
+	defer s.lk.Unlock()
+
+	groups := make(map[bitstr.Key][]mh.Multihash)
+	for _, h := range keys {
+		bs := bitstr.Key(key.BitString(keyspace.MhToBit256(h)))
+		p := bitstr.Key(bs[:s.prefixLen])
+		groups[p] = append(groups[p], h)
+	}
+
+	for prefix, toDel := range groups {
+		dsKey := s.dsKey(prefix)
+		data, err := s.ds.Get(ctx, dsKey)
```

Comment:
Is it necessary to Get before Delete? Seems better to just delete whether or not item is in datastore.

Reply:
Yes, we need to get the group of keys to which the target belong, remove it from the group, and write back the group to the datastore.